src/index.ml

```diff
@@ -417,39 +418,27 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
               l "[%s] found in %s" (Filename.basename t.root) name);
           ans
     in
+    let ( @~ ) a b = try a () with Not_found -> b () in
     let find_log_index () =
-      try find_if_exists ~name:"log" ~find:(fun log -> Tbl.find log.mem) t.log
-      with Not_found -> (
-        match
-          find_if_exists ~name:"index" ~find:interpolation_search t.index
-        with
-        | Some e -> e
-        | None -> raise Not_found )
+      find_if_exists ~name:"log" ~find:(fun log -> Tbl.find log.mem) t.log
+      @~ find_if_exists ~name:"index" ~find:interpolation_search t.index
     in
     IO.Mutex.with_lock t.rename_lock (fun () ->
         if t.config.readonly then sync_log t;
-        try
-          find_if_exists ~name:"log" ~find:(fun log -> Tbl.find log.mem) t.log
-        with Not_found -> (
-          try
-            find_if_exists ~name:"log_async"
-              ~find:(fun log -> Tbl.find log.mem)
-              t.log_async
-          with Not_found -> (
-            match
-              find_if_exists ~name:"index" ~find:interpolation_search t.index
-            with
-            | Some e -> e
-            | None when t.config.readonly ->
-                sync_log t;
-                find_log_index ()
-            | None -> raise Not_found ) ))
+        find_if_exists ~name:"log_async"
+          ~find:(fun log -> Tbl.find log.mem)
+          t.log_async
+        @~ fun () ->
+        find_log_index
+        @~
+        if t.config.readonly then (
```

Comment:
Doesn't this whole `if` need to be inside the `fun () ->`, else you're doing `sync_log` even if `find_log_index` returns the value?